### PR TITLE
Support assignment publishing

### PR DESCRIPTION
**Pull request by mstobo-coursera**
cc @ngarg-coursera

### CODE REVIEW

File: courseraprogramming/commands/publish.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+
+class ValidationError(Exception):
+    pass
+
+
+class InternalError(Exception):
+    pass
+
+
+def write_std_out(string, args):
```
Comment (saeta) [inline courseraprogramming/commands/publish.py]: Why not just use the logging facilities?

File: courseraprogramming/commands/publish.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+        '{}/{}~{}'.format(get_endpoint, course_id, item_id),
+        auth=auth)
+    if resp.status_code == 404:
+        raise ItemNotFoundError(course_id, item_id)
+    elif resp.status_code == 500:
+        raise InternalError()
+    return resp.json()['elements'][0]['metadata']
+
+
+def post_publish(
```
Comment (saeta) [inline courseraprogramming/commands/publish.py]: `post_publish` is not the most descriptive name. Could you either use a better name, or add a docstring to describe what the purpose of this is?

File: courseraprogramming/commands/publish.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+        help='The id of the course containing the assignment to publish.')
+
+    parser_publish.add_argument(
+        'item',
+        help='The id of the assignment to publish.')
+
+    parser_publish.add_argument(
+        '--additional_item',
+        action='append',
+        help='The next two args specify an item ID which will also be '
```
Comment (saeta) [inline courseraprogramming/commands/publish.py:198]: Do you need to specify `nargs`?

File: tests/commands/publish_tests.py [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from courseraprogramming.commands import publish
+
+from mock import ANY, call, MagicMock, patch
+
+
+class PublishParams:
```
Comment (saeta) [inline tests/commands/publish_tests.py:26]: Previously, you included a trailing `()` after the class name. Here you elide them. Can you pick one and be consistent? (Unless it's a pep8 violation, I'd prefer how you do it here.)

File: tests/commands/publish_tests.py [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from courseraprogramming.commands import publish
+
+from mock import ANY, call, MagicMock, patch
```
Comment (saeta) [inline tests/commands/publish_tests.py]: One per line, please. (Makes patches adding or removing one much easier to read.)

File: tests/commands/publish_tests.py [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+
+
+def fake_get_metadata(oauth, get_endpoint, course_id, item_id):
+    return {
+        "version": 0,
+        "authorId": 1234,
+        "publishedAssignmentId": 4321,
+        "updatedAt": 10000000,
+    }
+
```
Comment (saeta) [inline tests/commands/publish_tests.py:46]: Could you add a parser test to make sure command line parsing works how you expect it to?

File: courseraprogramming/commands/publish.py [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+
+import logging
+import requests
+import sys
+
+from courseraprogramming.commands import common
+from courseraprogramming.commands import oauth2
+
+
+class ErrorCodes():
```
Comment (saeta) [inline courseraprogramming/commands/publish.py]: Can you document that these are the program exit codes?

File: tests/commands/publish_tests.py (hunk 6 above)
Comment (mstobo-coursera) [inline tests/commands/publish_tests.py:46]: Added at the bottom

File: courseraprogramming/commands/__init__.py [hunk 8]
```diff
@@ -9,6 +9,7 @@
     "sanity",
     "upload",
     "version",
+    "publish",
```
Comment (saeta) [inline courseraprogramming/commands/__init__.py]: alpha sort please.

File: courseraprogramming/commands/publish.py [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+        'course',
+        help='The id of the course containing the assignment to publish.')
+
+    parser_publish.add_argument(
+        'item',
+        help='The id of the assignment to publish.')
+
+    parser_publish.add_argument(
+        '--additional_items',
+        nargs='*',
```
Comment (saeta) [inline courseraprogramming/commands/publish.py]: `s/nargs='*',/nargs='+',/`

**Comment by saeta** [comment]
Almost there. Just a couple nits. Otherwise LGTM.

**Comment by saeta** [comment]
LGTM. :shipit: :ship:

**Pull request merged**